### DISCUSSION TITLE
Just split inside a docker container

### DISCUSSION THREAD

**Discussion by SalahEddineGhamri** [wezterm/wezterm#5510] [General]
I like this terminal, font rendering is just fantastic and neovim feel smooth in it. I never liked tmux due to its strange keybindings. I like wezterm for its simplicity but I am a bit stuck on how to do a pane split while in a docker container. I want the pane to be spawn inside the container similar as if I "docker exec -it container /bin/bash" and run "tmux" within. is there a clean way to do it ? am like detecting which container id then spawning an exec domain in it ? 

I could follow the documentation to run the container in a tab using the menu launcher but nothing so far for the vertical and the horizontal splits 

```lua
function docker_list()
    local docker_list = {}
    local command = "docker container ls --format '{{.ID}}:{{.Names}}'"
    local handle = io.popen(command)
    if handle then
        for line in handle:lines() do
            local id, name = line:match("(.-):(.+)")
            if id and name then
                docker_list[id] = name
            end
        end
        handle:close()
    end
    return docker_list
end

function make_docker_label_func(id)
  return function(name)
    local command = "docker inspect --format '{{.State.Running}}' " .. id
    local handle = io.popen(command)
    local running = false
    if handle then
      local status = handle:read("*line")
      if status == "true" then
        running = true
      end
      handle:close()
    end
    local color = running and "Green" or "Red"
    return wezterm.format {
      { Foreground = { AnsiColor = color } },
      { Text = 'docker container named ' .. name },
    }
  end
end

function make_docker_fixup_func(id)
  return function(cmd)
    cmd.args = cmd.args or { '/bin/bash' }
    local wrapped = {
      'docker',
      'exec',
      '-it',
      id,
    }
    for _, arg in ipairs(cmd.args) do
      table.insert(wrapped, arg)
    end

    cmd.args = wrapped
    return cmd
  end
end

local exec_domains = {}
for id, name in pairs(docker_list()) do
  wezterm.log_info("ID: " .. id .. ", Name Salah: " .. name)
  table.insert(
    exec_domains,
    wezterm.exec_domain(
      'docker: ' .. name,
      make_docker_fixup_func(id),
      make_docker_label_func(id)
    )
  )
end

for k, v in pairs(exec_domains) do
  wezterm.log_info(k .. ": " .. tostring(v))
end

config.exec_domains = exec_domains
```

**Comment by wez** [comment]
FWIW, splits work OK for me here.  I slightly modified your example so that it works for an alpine container, which has `/bin/sh` but not `/bin/bash`, and to avoid using `io:popen`, which can block the GUI thread and which should be avoided. I also removed a space from the domain name generated for the docker containers, and made it into something that can be required as a standalone module:

File:  `~/.config/wezterm/docker.lua`:

```lua
local wezterm = require("wezterm")
local mod = {}

function docker_list()
	local docker_list = {}
	local success, stdout, stderr =
		wezterm.run_child_process({ "docker", "container", "ls", "--format", "{{.ID}}:{{.Names}}" })
	for _, line in ipairs(wezterm.split_by_newlines(stdout)) do
		local id, name = line:match("(.-):(.+)")
		if id and name then
			docker_list[id] = name
		end
	end
	return docker_list
end

function make_docker_label_func(id)
	return function(name)
		local success, stdout, stderr =
			wezterm.run_child_process({ "docker", "inspect", "--format", "{{.State.Running}}", id })
		local running = stdout == "true\n"
		local color = running and "Green" or "Red"
		return wezterm.format({
			{ Foreground = { AnsiColor = color } },
			{ Text = "docker container named " .. name },
		})
	end
end

function make_docker_fixup_func(id)
	return function(cmd)
		cmd.args = cmd.args or { "/bin/sh" }
		local wrapped = {
			"docker",
			"exec",
			"-it",
			id,
		}
		for _, arg in ipairs(cmd.args) do
			table.insert(wrapped, arg)
		end

		cmd.args = wrapped

		wezterm.log_info(wezterm.serde.json_encode(cmd))
		return cmd
	end
end

function mod.apply_to_config(config)
	local exec_domains = {}
	for id, name in pairs(docker_list()) do
		table.insert(
			exec_domains,
			wezterm.exec_domain("docker:" .. name, make_docker_fixup_func(id), make_docker_label_func(id))
		)
	end
	config.exec_domains = exec_domains
end

return mod
```

Then in my main `~/.wezterm.lua` I have:

```lua
local wezterm = require("wezterm")
local docker = require("docker")
local config = wezterm.config_builder()
docker.apply_to_config(config)
return config
```

With this I can spawn tabs into the alpine container and split them.

**Reply by SalahEddineGhamri** [reply]
Yes, thanks it works I have just tried. it is not what I have expected, the logic of workflow needs to be changed a bit now.

So the workflow with wezterm:
1. launch the wezterm with the configuration above
2. launch the docker container
3. "one restart of wezterm" here is needed to populate the launch menu (maybe something is missing from config above)
4. open a new tab with its ExecDomain from launch menu.
 
with tmux:
1. launch the terminal
2. launch the docker container
3. "docker exec -it" the container
5. launch tmux

However, with tmux steps 2-3-4 can be automazied via a shell script or python. with one command we are in tmux inside the container ready to play. 

I wonder, how to use wezterm maybe CLI to automate something that will bring me directly into ExecDomain ready to bang with splits and tabs and so on ?

**Comment by qlibp** [comment]
Method mentioned above should work with local container, but not remote container.
Any idea on how to make split on remote docker container directly?
